New "bad" assembly loader/remapper trips on MSBuild build

We haven't been able to compile MSBuild since https://github.com/mono/mono/commit/9ad5255338386a22e7a861e8bcc7c195dc591b99 was merged.

During compilation, MSBuild uses `AssemblyName.GetAssemblyName(sourcePath)?.Version` a bunch, which according to the documentation, "This method causes the file to be opened and closed, but the assembly is not added to this domain."

It seems our implementation loads anyway.

However, the new logic for remapping problematic Windows-only assemblies isn't triggered in this case.

Good:

```
Mono: Loading reference 7 of /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/msbuild-15.problem/artifacts/mono-msbuild/Microsoft.Build.dll asmctx DEFAULT, looking for System.IO.Compression, Version=4.1.2.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Mono: The request to load the assembly System.IO.Compression v4.1.2.0 was remapped to v4.0.0.0
Mono: Found assembly remapping for System.IO.Compression and was for the same version 4.0.0.0
Mono: Domain MSBuild.dll search path is:
Mono:   path[0] = '/Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/msbuild-15.problem/artifacts/mono-msbuild/'
Mono: End of domain MSBuild.dll search path.
Mono: Assembly Loader probing location: '/Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/msbuild-15.problem/artifacts/mono-msbuild/System.IO.Compression.dll'.
Mono: Denying load of problematic image /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/msbuild-15.problem/artifacts/mono-msbuild/System.IO.Compression.dll
Mono: Unloading image /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/msbuild-15.problem/artifacts/mono-msbuild/System.IO.Compression.dll [0x7f94dd1db000].
Mono: Assembly Loader probing location: '/Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/stage/lib/mono/gac/System.IO.Compression/4.0.0.0__b77a5c561934e089/System.IO.Compression.dll'.
Mono: Image addref System.IO.Compression[0x7f94dc52f960] (asmctx DEFAULT) -> /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/stage/lib/mono/gac/System.IO.Compression/4.0.0.0__b77a5c561934e089/System.IO.Compression.dll[0x7f94dd1db000]: 2
Mono: Prepared to set up assembly 'System.IO.Compression' (/Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/stage/lib/mono/gac/System.IO.Compression/4.0.0.0__b77a5c561934e089/System.IO.Compression.dll)
Mono: Assembly System.IO.Compression[0x7f94dc52f960] added to domain MSBuild.dll, ref_count=1
Mono: Assembly Loader loaded assembly from location: '/Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/stage/lib/mono/gac/System.IO.Compression/4.0.0.0__b77a5c561934e089/System.IO.Compression.dll'.
Mono: Assembly Ref addref Microsoft.Build[0x7f94dc6168d0] -> System.IO.Compression[0x7f94dc52f960]: 2
```

Bad:
```
Mono: Assembly Loader probing location: '/Users/builder/.nuget/packages/system.io.compression/4.3.0/ref/net46/System.IO.Compression.dll'.
Mono: Image addref System.IO.Compression[0x7f94def4b060] (asmctx REFONLY) -> /Users/builder/.nuget/packages/system.io.compression/4.3.0/ref/net46/System.IO.Compression.dll[0x7f94df100200]: 2
Mono: Prepared to set up assembly 'System.IO.Compression' (/Users/builder/.nuget/packages/system.io.compression/4.3.0/ref/net46/System.IO.Compression.dll)
Mono: Assembly System.IO.Compression[0x7f94def4b060] added to domain MSBuild.dll, ref_count=1
Mono: Assembly Loader loaded assembly from location: '/Users/builder/.nuget/packages/system.io.compression/4.3.0/ref/net46/System.IO.Compression.dll'.
```

The end result, during build:
```
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018: The "ResolvePackageFileConflicts" task failed unexpectedly. [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018: This is an unhandled exception from a task -- PLEASE OPEN A BUG AGAINST THE TASK OWNER. [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018: System.IO.FileNotFoundException:  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018: File name: '/Users/builder/.nuget/packages/system.io.compression/4.3.0/ref/net46/System.IO.Compression.dll' [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at (wrapper managed-to-native) System.Reflection.Assembly.InternalGetAssemblyName(string,Mono.MonoAssemblyName&,string&) [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at System.Reflection.AssemblyName.GetAssemblyName (System.String assemblyFile) [0x0001a] in <0a1489687ea248b2a706d89d4c87bbcf>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.FileUtilities.GetAssemblyVersion (System.String sourcePath) [0x00000] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.FileUtilities.TryGetAssemblyVersion (System.String sourcePath) [0x00016] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.ConflictResolution.ConflictItem.get_AssemblyVersion () [0x0004e] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.ConflictResolution.ConflictResolver`1[TConflictItem].ResolveConflict (TConflictItem item1, TConflictItem item2) [0x000c2] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.ConflictResolution.ConflictResolver`1[TConflictItem].ResolveConflicts (System.Collections.Generic.IEnumerable`1[T] conflictItems, System.Func`2[T,TResult] getItemKey, System.Action`1[T] foundConflict, System.Boolean commitWinner, System.Action`1[T] unresolvedConflict) [0x0003a] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.ConflictResolution.ResolvePackageFileConflicts.ExecuteCore () [0x000a6] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.NET.Build.Tasks.TaskBase.Execute () [0x00000] in <a39f5cd91a004fcf87657c269fcb8060>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x0002a] in <143af2be99a94a8480f7f49f5d47377c>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
15:43:22 @/external/bockbuild/builds/msbuild-15/artifacts/mono-msbuild/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ConflictResolution.targets(41,5): error MSB4018:   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002a2] in <143af2be99a94a8480f7f49f5d47377c>:0  [@/external/bockbuild/builds/msbuild-15/src/NuGetSdkResolver.UnitTests/NuGet.MSBuildSdkResolver.UnitTests.csproj]
```

--- a/mono/metadata/icall.c
+++ b/mono/metadata/icall.c
@@ -101,6 +101,7 @@
 #include <mono/metadata/w32error.h>
 #include <mono/utils/w32api.h>
 #include <mono/utils/mono-merp.h>
+#include <mono/utils/mono-logger-internals.h>
 
 #include "decimal-ms.h"
 #include "number-ms.h"
@@ -5447,6 +5448,8 @@ ves_icall_System_Reflection_Assembly_InternalGetAssemblyName (MonoStringHandle f
 	dirname = g_path_get_dirname (filename);
 	replace_shadow_path (mono_domain_get (), dirname, &filename);
 	g_free (dirname);
+
+	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_ASSEMBLY, "InternalGetAssemblyName (\"%s\")", filename);
 
 	image = mono_image_open_full (filename, &status, TRUE);
 

--- a/mono/metadata/image.c
+++ b/mono/metadata/image.c
@@ -1552,6 +1552,9 @@ register_image (MonoImage *image)
 		g_hash_table_insert (loaded_images_by_name, (char *) image->assembly_name, image);
 	mono_images_unlock ();
 
+	if (mono_is_problematic_image (image)) {
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_ASSEMBLY, "Registering %s, problematic image '%s'", image->ref_only ? "REFONLY" : "default", image->name);
+	}
 	return image;
 }
 
@@ -1773,6 +1776,7 @@ mono_image_open_a_lot (const char *fname, MonoImageOpenStatus *status, gboolean 
 			//  to see it again when we go searching for an image
 			//  to load.
 			mono_images_unlock ();
+			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_ASSEMBLY, "Not returning problematic image '%s' refcount=%d", image->name, image->ref_count);
 			return NULL;
 		}
 		mono_image_addref (image);

--- a/mono/metadata/image.c
+++ b/mono/metadata/image.c
@@ -1763,7 +1763,7 @@ mono_image_open_a_lot (const char *fname, MonoImageOpenStatus *status, gboolean 
 	g_free (absfname);
 
 	if (image) { // Image already loaded
-		if (!load_from_context && mono_is_problematic_image (image)) {
+		if (!refonly && !load_from_context && mono_is_problematic_image (image)) {
 			// If we previously loaded a problematic image, don't
 			// return it if we're not in LoadFrom context.
 			//

--- a/mono/mini/method-to-ir.c
+++ b/mono/mini/method-to-ir.c
@@ -11486,7 +11486,8 @@ mono_ldptr:
 						}
 					}
 
-					if ((invoke_context_used == 0 || !cfg->gsharedvt) || cfg->llvm_only) {
+					/* FIXME: SGEN support */
+					if (invoke_context_used == 0 || cfg->llvm_only) {
 						if (cfg->verbose_level > 3)
 							g_print ("converting (in B%d: stack: %d) %s", cfg->cbb->block_num, (int)(sp - stack_start), mono_disasm_code_one (NULL, method, ip + 6, NULL));
 						if ((handle_ins = handle_delegate_ctor (cfg, ctor_method->klass, target_ins, cmethod, context_used, FALSE))) {
@@ -11547,7 +11548,8 @@ mono_ldptr:
 					if (mono_security_core_clr_enabled ())
 						ensure_method_is_allowed_to_call_method (cfg, method, ctor_method);
 
-					if ((invoke_context_used == 0 || !cfg->gsharedvt) || cfg->llvm_only) {
+					/* FIXME: SGEN support */
+					if (invoke_context_used == 0 || cfg->llvm_only) {
 						if (cfg->verbose_level > 3)
 							g_print ("converting (in B%d: stack: %d) %s", cfg->cbb->block_num, (int)(sp - stack_start), mono_disasm_code_one (NULL, method, ip + 6, NULL));
 						if ((handle_ins = handle_delegate_ctor (cfg, ctor_method->klass, target_ins, cmethod, context_used, is_virtual))) {

--- a/mono/mini/mini-runtime.c
+++ b/mono/mini/mini-runtime.c
@@ -2376,7 +2376,9 @@ lookup_start:
 
 	if (!code) {
 		if (mono_class_is_open_constructed_type (m_class_get_byval_arg (method->klass))) {
-			mono_error_set_invalid_operation (error, "Could not execute the method because the containing type is not fully instantiated.");
+			char *full_name = mono_type_get_full_name (method->klass);
+			mono_error_set_invalid_operation (error, "Could not execute the method because the containing type '%s', is not fully instantiated.", full_name);
+			g_free (full_name);
 			return NULL;
 		}
 

